### PR TITLE
cmds/core/wget: remove testutil

### PR DESCRIPTION
**Pull request by binjip978**
Signed-off-by: Siarhiej Siemianczuk <pdp.eleven11@gmail.com>

### CODE REVIEW

File: cmds/core/wget/wget.go [hunk 1]
```diff
@@ -45,16 +45,16 @@ func usage() {
...
+func run(args []string) (reterr error) {
 	log.SetPrefix("wget: ")
 
-	if flag.Parse(); flag.NArg() != 1 {
+	if len(args) != 1 {
 		usage()
 	}
 
-	argURL := flag.Arg(0)
+	argURL := args[0]
```
Comment (rminnich) [inline cmds/core/wget/wget.go:55]: does this work if args is nil?

File: cmds/core/wget/wget.go [hunk 2]
```diff
@@ -91,8 +87,16 @@ func run() (reterr error) {
 	return nil
 }
 
+func defaultOutputPath(url *url.URL) string {
+	if url.Path != "" && url.Path[len(url.Path)-1] != '/' {
```
Comment (rminnich) [inline cmds/core/wget/wget.go:91]: would strings.Suffix work here? 
I think it would and it makes this simpler
https://go.dev/play/p/7TyYCf1DmRe

**Review by rminnich — COMMENTED** [review]
Just two questions, nice to see more tests!

File: cmds/core/wget/wget.go (hunk 1 above)
Comment (rminnich) [reply cmds/core/wget/wget.go:55]: or if len(args) is 0?

Comment (binjip978) [reply cmds/core/wget/wget.go:56]: Yes, args can't be nil, we check it length on line 51, so it's only check if someone send `wget ""`. It's probably can be improved in url parsing.

I will try to address it in next PR, at least wget has a bug
```sh
./wget localhost:8080/200
wget: 2022/10/28 07:39:37 failed to download localhost:8080/200: encountered error no such scheme with "localhost:8080/200"
```

File: cmds/core/wget/wget.go (hunk 2 above)
Comment (binjip978) [reply cmds/core/wget/wget.go]: Updated.

File: cmds/core/wget/wget.go [hunk 3]
```diff
@@ -45,16 +46,16 @@ func usage() {
 	os.Exit(2)
 }
 
-func run() (reterr error) {
+func run(args []string) (reterr error) {
 	log.SetPrefix("wget: ")
 
-	if flag.Parse(); flag.NArg() != 1 {
+	if len(args) != 1 {
```
Comment (rminnich) [inline cmds/core/wget/wget.go:52]: if len(args) has to be 1, I think this should be run(url string) error
Otherwise, we are letting coverage force us into writing strange code. It's pretty nonintuitive (to me) to have an array as a parameter that can only be one element.
Your other option is to call run(flag.Arg(0)) in main (https://pkg.go.dev/flag#Arg) which returns "" if the numbered arg does not exist.
If there are no args, flag.Arg(0) will be "", and that will fall through to your error message test on line 57.

Comment (binjip978) [reply cmds/core/wget/wget.go]: Sure, this make sense. flag.Args(0) will look much better.

File: cmds/core/wget/wget.go [hunk 4]
```diff
@@ -45,29 +46,20 @@ func usage() {
...
 }
 
-func run() (reterr error) {
+func run(arg string) (reterr error) {
 	log.SetPrefix("wget: ")
 
-	if flag.Parse(); flag.NArg() != 1 {
-		usage()
+	if arg == "" {
+		return errors.New("empty URL")
```
Comment (rminnich) [inline cmds/core/wget/wget.go:53]: If you create an error:
var ErrEmptyURL = errors.new("empty URL")
then you can test for that error in your tests, with errors.Is()
Also, no need to name the return parrameter, I think; all your returns provide the error result, so
func run(arg string) error {

File: cmds/core/wget/wget_test.go [hunk 5]
```diff
@@ -165,6 +159,22 @@ func TestWget(t *testing.T) {
 	}
 }
 
-func TestMain(m *testing.M) {
-	testutil.Run(m, main)
+func TestDefaultOutputPath(t *testing.T) {
+	u1, _ := url.Parse("http://localhost:8080/")
```
Comment (abrender) [inline cmds/core/wget/wget_test.go:163]: Ignoring errors returned by `url.Parse` seems a bit dangerous.   Maybe it works today but someone may add an invalid url in the future and this may get missed.

Given that you have 3 different test cases here maybe it makes sense to change this to a table-driven test design and avoid having to duplicate the url parsing and error checking for each individual case?

File: cmds/core/wget/wget.go [hunk 6]
```diff
@@ -45,29 +46,20 @@ func usage() {
...
 		return err
 	}
 
 	if *outPath == "" {
-		if url.Path != "" && url.Path[len(url.Path)-1] != '/' {
-			*outPath = path.Base(url.Path)
-		} else {
-			*outPath = "index.html"
-		}
+		*outPath = defaultOutputPath(parsedURL)
```
[the remark this answers is not in the document]
Comment (binjip978) [reply cmds/core/wget/wget.go:62]: Initial intent was to just remove deprecated testutil. But looks like I will rewrite it a bit :), the is at least one bug with not using default http scheme.

Probably I reuse command struct similar to tee util.

File: cmds/core/wget/wget_test.go (hunk 5 above)
Comment (binjip978) [reply cmds/core/wget/wget_test.go]: Actually we don't need full url to infer output path so I remove all url parsing, and this also allow to use table driven tests.

File: cmds/core/wget/wget.go (hunk 4 above)
Comment (binjip978) [reply cmds/core/wget/wget.go]: Updated, actually using errors.Is will not work, because HTTPClientCodeError has nil error when status is not 200.

File: cmds/core/wget/wget.go [hunk 7]
```diff
@@ -32,42 +32,41 @@ import (
...
+	url        string
+	outputPath string
 }
 
-func run() (reterr error) {
-	log.SetPrefix("wget: ")
-
-	if flag.Parse(); flag.NArg() != 1 {
-		usage()
+func New(outPath string, url string) *command {
```
Comment (rminnich) [inline cmds/core/wget/wget.go:49]: this can be new, not New, otherwise some of the code checkers will complain, and we don't need to export it.

**Review by rminnich — COMMENTED** [review]
I like this new design. You can see my one comment re New() vs new()
Let's see what abrender thinks but I believe we are almost done!

File: cmds/core/wget/wget.go (hunk 7 above)
Comment (binjip978) [reply cmds/core/wget/wget.go:49]: Yeah sure, but I make it New because otherwise code checkers may or will complain for shadowing builtin function, this is not an issue here, right now and probably never will, I don't expect someone will want to call builtin. That why I initially named it `tee`, but probably `newCommand` or `newUtil` will be a compromise?

File: cmds/core/wget/wget_test.go [hunk 8]
```diff
@@ -165,6 +156,29 @@ func TestWget(t *testing.T) {
 	}
 }
 
-func TestMain(m *testing.M) {
-	testutil.Run(m, main)
+func TestDefaultOutputPath(t *testing.T) {
+	tests := []struct {
+		path   string
+		output string
```
Comment (abrender) [inline cmds/core/wget/wget_test.go:162]: nit: it's common to name/prefix these with `want`, so that it's clear to someone reading/writing tests which test case fields are the values you're setting vs which are the values you're expecting.

File: cmds/core/wget/wget_test.go [hunk 9]
```diff
@@ -40,120 +43,108 @@ func (h handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
-		t.Fatalf("error setting up TCP listener: %v", err)
+		return 0, err
 	}
-	return l, l.Addr().(*net.TCPAddr).Port
+	l, err := net.ListenTCP("tcp", a)
+	if err != nil {
+		return 0, err
+	}
+
+	l.Close()
```
Comment (abrender) [inline cmds/core/wget/wget_test.go:56]: IIUC this may not work as intended. l.Close() will close the listening port, so the port number *may* be reused by another test/process (in a race condition).   The old code relies on the caller of `getUnusedPort` to defer the port closing when the test is done... and it also sets up a fake TCP listener on that port which immediately closes the connection (you'll probably have to change the test to expect something other than `syscall.ECONNREFUSED`).

I think it makes sense to copy that here, WDYT?

File: cmds/core/wget/wget_test.go [hunk 10]
```diff
@@ -40,120 +43,108 @@ func (h handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+		{
+			name:       "ipv4",
+			url:        fmt.Sprintf("%s/200", srv.URL),
+			content:    content,
+			outputPath: filepath.Join(tmpDir, "basic"),
+			err:        nil,
+		},
+		{
+			name:       "localhost",
+			url:        strings.Replace(srv.URL, "127.0.0.1", "localhost", 1) + "/200",
```
Comment (abrender) [inline cmds/core/wget/wget_test.go:86]: this (`127.0.0.1` -> localhost) will break when we change to IPv4 right? Is there any value to this test? What `localhost` means is specific to the OS, DNS resolvers / `hosts` file, etc right? Maybe we should get rid of the 'localhost' tests?

File: cmds/core/wget/wget.go [hunk 11]
```diff
@@ -79,20 +78,37 @@ func run() (reterr error) {
...
-	reader, err := schemes.FetchWithoutCache(context.Background(), url)
+	reader, err := schemes.FetchWithoutCache(context.Background(), parsedURL)
 	if err != nil {
-		return fmt.Errorf("Failed to download %v: %v", argURL, err)
+		return fmt.Errorf("failed to download %v: %w", c.url, err)
 	}
 
-	if err := uio.ReadIntoFile(reader, *outPath); err != nil {
+	if err := uio.ReadIntoFile(reader, c.outputPath); err != nil {
 		return err
```
[the remark this answers is not in the document]
Comment (binjip978) [reply cmds/core/wget/wget.go:87]: Not sure about all errors, but I may lead to path duplication, here how it looks if user don't have permissions:
`wget: 2022/11/09 08:24:11 open /Users/binjip978/src/u-root/u-root/cmds/core/wget/aaa: permission denied`

So I think it's better to leave it as is.

File: cmds/core/wget/wget_test.go (hunk 9 above)
Comment (binjip978) [reply cmds/core/wget/wget_test.go]: You are absolutely right, ports can be reused. I probably was confused by test name "no server". So I revert to previous approach of closing connections and checking for io.EOF error.

File: cmds/core/wget/wget_test.go (hunk 10 above)
Comment (binjip978) [reply cmds/core/wget/wget_test.go]: Yeah, I'm not quite getting what this test was trying to accomplish, but it was there before so probably original author know more, so to be on the save side it probably will not harm if it will stay for a while.

File: cmds/core/wget/wget_test.go [hunk 12]
```diff
@@ -40,131 +43,147 @@ func (h handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+	tmpDir := t.TempDir()
+	freePort, closer := unusedPort(t)
+	defer closer()
+
+	var tests = []struct {
+		name        string
+		url         string // in
+		wantContent string // out
+		outputPath  string
+		err         error
```
Comment (abrender) [inline cmds/core/wget/wget_test.go:85]: can you please name this `wantErr`, to distinguish fields that set values in the test case, vs fields that are the expected outputs? Thanks!

File: cmds/core/wget/wget.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (abrender) [reply cmds/core/wget/wget.go:87]: Ah, I didn't realize that the error already returned the filename - thanks for pointing that out!

File: cmds/core/wget/wget_test.go [hunk 13]
```diff
@@ -40,131 +43,147 @@ func (h handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
-		url:     "http://localhost:%[2]d/200",
-		content: "",
-		retCode: 1,
-	},
-}
-
-func getListener(t *testing.T) (net.Listener, int) {
+// unusedPort return a port number and setup a listener that closes connection
+// for any request
+func unusedPort(t *testing.T) (int, func() error) {
```
Comment (abrender) [inline cmds/core/wget/wget_test.go:48]: Now that the code is actually listening and not returning an "unused port" - can you think of a better name & docstring for this function? Thanks!

File: cmds/core/wget/wget_test.go [hunk 14]
```diff
@@ -40,131 +43,147 @@ func (h handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+					t.Errorf("wanted:\n%#v\ngot:\n%#v", tt.wantContent, string(content))
 				}
 			}
 		})
 	}
 }
 
-func TestMain(m *testing.M) {
-	testutil.Run(m, main)
+func TestDefaultOutputPath(t *testing.T) {
```
Comment (abrender) [inline cmds/core/wget/wget_test.go:164]: Can this be integrated into TestWget?  Each test case already has `outputPath` to set the flag,  maybe you can add `wantOutputPath` and update the test case to read that file?       That would test the end-to-end functionality of wget.

TestDefaultOutputPath is testing an implementation detail - if one day in the future we wanted to rename the function or use some other way to determine the output path then we'd have to update this test code, even though the user's experience wouldn't change.   If you test this functionality as part of TestWget then we can change the implementation in the future and not have to update the tests.

Thanks!

File: cmds/core/wget/wget_test.go [hunk 15]
```diff
@@ -40,131 +43,147 @@ func (h handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+// for any request
+func unusedPort(t *testing.T) (int, func() error) {
 	t.Helper()
-	l, err := net.Listen("tcp", ":0")
+	a, err := net.ResolveTCPAddr("tcp", ":0")
 	if err != nil {
-		t.Fatalf("error setting up TCP listener: %v", err)
+		t.Fatal(err)
+	}
+	l, err := net.ListenTCP("tcp", a)
```
Comment (abrender) [inline cmds/core/wget/wget_test.go:54]: I see you've split out `net.Listen` into `net.ResolveTCPAddr` and `net.ListenTCP`, can you expand on why you chose to do that?   I  am not too familiar with the `net` module so perhaps you are changing this for a reason that I do not understand. Thanks!

File: cmds/core/wget/wget_test.go [hunk 16]
```diff
@@ -40,131 +43,147 @@ func (h handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+		url         string // in
+		wantContent string // out
+		outputPath  string
+		err         error
+	}{
+		{
+			name:        "ipv4",
+			url:         fmt.Sprintf("%s/200", srv.URL),
+			wantContent: content,
+			outputPath:  filepath.Join(tmpDir, "basic"),
```
Comment (abrender) [inline cmds/core/wget/wget_test.go:91]: instead of doing `filepath.Join(tmpDir, ...)` in each test case, maybe you can move the `tmpDir := t.TempDir()` and the `filepath.Join` into the test loop code?

File: pkg/curl/schemes.go [hunk 17]
```diff
@@ -395,7 +396,7 @@ func httpFetch(ctx context.Context, c *http.Client, u *url.URL) (io.Reader, erro
 	}
 
 	if resp.StatusCode != 200 {
-		return nil, &HTTPClientCodeError{err, resp.StatusCode}
+		return nil, &HTTPClientCodeError{ErrStatusNotOk, resp.StatusCode}
```
Comment (abrender) [inline pkg/curl/schemes.go:399]: is this change needed? Isn't `ErrStatusNotOk` redundant with `resp.StatusCode != 200`, which can be determined by the caller of this function?  This changes the behavior such that the returned `HTTPClientCodeErr.Err` field is no longer `nil` which may not be backwards compatible - are we sure this is safe?

**Review by abrender — COMMENTED** [review]
Thanks for continuing to work on this - it is appreciated!

File: pkg/curl/schemes.go (hunk 17 above)
Comment (binjip978) [reply pkg/curl/schemes.go:399]: it's needed if we want to use `errors.Is`. Our curl package treat any non 200 response as an error. And because err in `&HTTPClientCodeError{err, resp.StatusCode}` is nil we can't unwrap it to an error.

It's not clear to me why any non 200 code is an error, but it's kinda different story.

File: cmds/core/wget/wget_test.go (hunk 15 above)
Comment (binjip978) [reply cmds/core/wget/wget_test.go]: Sure, when I got an idea about connection refused I just rewrite whole thing from scratch, and it make sense to revert original code back.

File: cmds/core/wget/wget_test.go (hunk 13 above)
Comment (binjip978) [reply cmds/core/wget/wget_test.go]: updated

File: cmds/core/wget/wget_test.go (hunk 12 above)
Comment (binjip978) [reply cmds/core/wget/wget_test.go]: yep

File: cmds/core/wget/wget_test.go (hunk 16 above)
Comment (binjip978) [reply cmds/core/wget/wget_test.go]: updated

**Comment by binjip978** [comment]
looks like we can't guarantee EOF `wget_test.go:143: expected: EOF, got: failed to download http://localhost:33527/200: encountered error Get "http://localhost:33527/200": http: server closed`

pervious test just check for non zero, but this seems flaky on CI.

File: cmds/core/wget/wget_test.go [hunk 18]
```diff
@@ -40,131 +43,144 @@ func (h handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+				t.Errorf("expected: %v, got: %v", tt.wantErr, err)
 			}
 
-			if tt.content != "" {
-				content, err := os.ReadFile(filepath.Join(tmpDir, fileName))
+			if tt.wantContent != "" {
+				content, err := os.ReadFile(tt.outputPath)
 				if err != nil {
-					t.Errorf("File %s was not created: %v", fileName, err)
+					t.Errorf("file %s was not created: %v", tt.outputPath, err)
```
Comment (abrender) [inline cmds/core/wget/wget_test.go:149]: t.Fatalf here?  If `err` is not `nil` then the check of "content" on line 153 won't make any sense.

File: cmds/core/wget/wget_test.go (hunk 14 above)
Comment (abrender) [reply cmds/core/wget/wget_test.go:161]: > index.html will depend from current directory and it don't work well with tempDir, so I don't think it worth to do.

I think this can be solved if you simply change into a tempdir: `os.Chdir(t.TempDir())`.   Here's a working example  that I came up with: https://gist.github.com/abrender/635144e451259825eb9e00527108205d

Comment (binjip978) [reply cmds/core/wget/wget_test.go:128]: Yes it can, but for one line of code that extracted and tested separately is't worth it? I mean I never made a chdir in test and it definitely will work here. But I'm afraid how it will affect other tests? Should I defer it back to original and what about defer and if we have Fatal? IMHO it's just became to complex.

**Comment by binjip978** [comment]
> looks like we can't guarantee EOF `wget_test.go:143: expected: EOF, got: failed to download http://localhost:33527/200: encountered error Get "http://localhost:33527/200": http: server closed`
> 
> pervious test just check for non zero, but this seems flaky on CI.

Can't made it work. It may return at least 3 different errors and one of the is even internal `errServerClosedIdle = errors.New("http: server closed idle connection")`. So I diced to remove this test, seems like it's not testing anything in wget command, and probably does not do anything pkg/curl either. Error is returned in httpFetch functions in scheme.go in line 393 `resp, err := c.Do(req)`.

Or I can remove errors.Is and check for error code 1 or 0.

File: cmds/core/wget/wget_test.go (hunk 14 above)
Comment (abrender) [reply cmds/core/wget/wget_test.go:128]: Go itself uses chdir while testing file / directory stuff (example: https://github.com/golang/go/blob/460fd63cccd2f1d16fc4b1b761545b1649e14e28/src/io/fs/walk_test.go#L102-L109).  Yes, good point about calling defer back to the original directory :)

I appreciate you questioning the trade off and complexity here. However I do believe that it's worth it here, ensuring that the correct output file gets written is a critical part of the program.  While testing `defaultOutputPath` is good, there's no guarantee that the function itself will be called or used correction - which is what this test covers.

**Comment by abrender** [comment]
> > looks like we can't guarantee EOF `wget_test.go:143: expected: EOF, got: failed to download http://localhost:33527/200: encountered error Get "http://localhost:33527/200": http: server closed`
> > pervious test just check for non zero, but this seems flaky on CI.
> 
> Can't made it work. It may return at least 3 different errors and one of the is even internal `errServerClosedIdle = errors.New("http: server closed idle connection")`. So I diced to remove this test, seems like it's not testing anything in wget command, and probably does not do anything pkg/curl either. Error is returned in httpFetch functions in scheme.go in line 393 `resp, err := c.Do(req)`.
> 
> Or I can remove errors.Is and check for error code 1 or 0.

Perhaps [cmpopts.EquateErrors](https://pkg.go.dev/github.com/google/go-cmp/cmp/cmpopts#pkg-variables) and [AnyError](https://pkg.go.dev/github.com/google/go-cmp/cmp/cmpopts#pkg-variables) can help here?

```
import(
    "github.com/google/go-cmp/cmp"
    "github.com/google/go-cmp/cmp/cmpopts"
)

// Update test case to have `err: cmpopts.AnyError` instead of io.EOF

if !cmp.Equal(tt.err, err, cmpopts.EquateErrors()) {
  t.Errorf("Got error=%v, expected=%v", err, tt.err)
}
if (err != nil || tt.err != nil ) {
   return
}
...



```

**Comment by binjip978** [comment]
> > > looks like we can't guarantee EOF `wget_test.go:143: expected: EOF, got: failed to download http://localhost:33527/200: encountered error Get "http://localhost:33527/200": http: server closed`
> > > pervious test just check for non zero, but this seems flaky on CI.
> > 
> > 
> > Can't made it work. It may return at least 3 different errors and one of the is even internal `errServerClosedIdle = errors.New("http: server closed idle connection")`. So I diced to remove this test, seems like it's not testing anything in wget command, and probably does not do anything pkg/curl either. Error is returned in httpFetch functions in scheme.go in line 393 `resp, err := c.Do(req)`.
> > Or I can remove errors.Is and check for error code 1 or 0.
> 
> Perhaps [cmpopts.EquateErrors](https://pkg.go.dev/github.com/google/go-cmp/cmp/cmpopts#EquateErrors) and [AnyError](https://pkg.go.dev/github.com/google/go-cmp/cmp/cmpopts#pkg-variables) can help here?
> 
> ```
> import(
>     "github.com/google/go-cmp/cmp"
>     "github.com/google/go-cmp/cmp/cmpopts"
> )
> 
> // Update test case to have `err: cmpopts.AnyError` instead of io.EOF
> 
> if !cmp.Equal(tt.err, err, cmpopts.EquateErrors()) {
>   t.Errorf("Got error=%v, expected=%v", err, tt.err)
> }
> if (err != nil || tt.err != nil ) {
>    return
> }
> ...
> ```

This will work, but I would prefer to just separate this test case and do simple `if err != nil { t.Error() }` and don't bring any additional dependencies. This is better aligned with go values "a little copying is better than a little dependency".

The problem with this that I don't fully understand what the purpose of this test case. We already covered error case for curl command with 4xx and 5xx test cases. What kind of http related error we got there it's not a wget command problem, it's maybe a problem of curl lib, but error comes from http client do function. 
But I probably missing something.

**Comment by abrender** [comment]
> > > > looks like we can't guarantee EOF `wget_test.go:143: expected: EOF, got: failed to download http://localhost:33527/200: encountered error Get "http://localhost:33527/200": http: server closed`
> > > > pervious test just check for non zero, but this seems flaky on CI.
> > > 
> > > 
> > > Can't made it work. It may return at least 3 different errors and one of the is even internal `errServerClosedIdle = errors.New("http: server closed idle connection")`. So I diced to remove this test, seems like it's not testing anything in wget command, and probably does not do anything pkg/curl either. Error is returned in httpFetch functions in scheme.go in line 393 `resp, err := c.Do(req)`.
> > > Or I can remove errors.Is and check for error code 1 or 0.
> > 
> > 
> > Perhaps [cmpopts.EquateErrors](https://pkg.go.dev/github.com/google/go-cmp/cmp/cmpopts#EquateErrors) and [AnyError](https://pkg.go.dev/github.com/google/go-cmp/cmp/cmpopts#pkg-variables) can help here?
> > ```
> > import(
> >     "github.com/google/go-cmp/cmp"
> >     "github.com/google/go-cmp/cmp/cmpopts"
> > )
> > 
> > // Update test case to have `err: cmpopts.AnyError` instead of io.EOF
> > 
> > if !cmp.Equal(tt.err, err, cmpopts.EquateErrors()) {
> >   t.Errorf("Got error=%v, expected=%v", err, tt.err)
> > }
> > if (err != nil || tt.err != nil ) {
> >    return
> > }
> > ...
> > ```
> 
> This will work, but I would prefer to just separate this test case and do simple `if err != nil { t.Error() }` and don't bring any additional dependencies. This is better aligned with go values "a little copying is better than a little dependency".
> 
> The problem with this that I don't fully understand what the purpose of this test case. We already covered error case for curl command with 4xx and 5xx test cases. What kind of http related error we got there it's not a wget command problem, it's maybe a problem of curl lib, but error comes from http client do function. But I probably missing something.

Sure, a separate test for `err != nil` sounds fine to me :) I was just throwing out a solution if you wanted to keep it in a single test.

As for what it's testing - I'm not the original author, but I assume there's a reasonable assumption that error handling for HTTP status codes is different than error handling for errors unrelated to HTTP status codes (like network errors).

File: cmds/core/wget/wget_test.go [hunk 19]
```diff
@@ -40,63 +44,122 @@ func (h handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tempDir := t.TempDir()
+			err := os.Chdir(tempDir)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			if tt.outputPath != "" {
+				tt.outputPath = filepath.Join(tempDir, tt.outputPath)
```
Comment (abrender) [inline cmds/core/wget/wget_test.go:136]: The test now `ChDir`s into tempDir so there's no need to do this `filepath.Join` right?

File: cmds/core/wget/wget_test.go [hunk 20]
```diff
@@ -40,63 +44,122 @@ func (h handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+			if tt.wantErr == nil && err != nil {
+				t.Fatalf("expected nil, got: %v", err)
+			}
+			if tt.wantErr != nil && !errors.Is(err, tt.wantErr) {
+				t.Fatalf("expected: %v, got: %v", tt.wantErr, err)
+			}
+
+			if tt.wantContent != "" {
+				if tt.outputPath == "" {
+					tt.outputPath = "./index.html"
```
Comment (abrender) [inline cmds/core/wget/wget_test.go:149]: Instead of hard coding this logic here, can you please add a `wantOutputPath` field to the test cases and be explicit about the results we expect, along with some test cases of sub directories, etc? I added a few ones that I thought of in https://gist.github.com/abrender/635144e451259825eb9e00527108205d ? Thanks!

Comment (binjip978) [reply cmds/core/wget/wget_test.go]: Added, thanks for other test cases!

**Comment by abrender** [comment]
Looks good to me! Thank you so much for your patience.

I'll await @rminnich to resolve his open comment threads

File: cmds/core/wget/wget_test.go (hunk 19 above)
Comment (binjip978) [reply cmds/core/wget/wget_test.go]: yep

**Comment by binjip978** [comment]
Sorry @rminnich but I can't change labels, so pinging you via mention :)

File: cmds/core/wget/wget.go [hunk 21]
```diff
@@ -32,42 +32,41 @@ import (
...
+	url        string
+	outputPath string
 }
 
-func run() (reterr error) {
-	log.SetPrefix("wget: ")
-
-	if flag.Parse(); flag.NArg() != 1 {
-		usage()
+func newCommand(outPath string, url string) *command {
```
Comment (rminnich) [inline cmds/core/wget/wget.go:49]: note that you can say 
output, url string
gofmt -s may or may not simplify this or complain about it ...

**Review by rminnich — APPROVED** [review]
This is a substantial improvement. Thanks for your patience.

**Review by rminnich — APPROVED** [review]

**Review by rminnich — APPROVED** [review]
